### ISSUE TITLE
Epic: Shadow DOM for all components

### ISSUE THREAD

**Issue by jcfranco** [Esri/calcite-components#2570]
Components should avoid using scoped styles as they are not ideal for the following reasons:

1. VDOM integration – VDOMs have their representation of the DOM tree that can potentially clash with any elements within the scoped component's 'internal' HTML structure
2. Breaks encapsulation 
   a. Component internal elements are exposed in light DOM queries   
   b. Scoped CSS is prone to being overridden with high-specificity selectors
   c. Scoped CSS can clash with external ones unless the component's selectors are namespaced (#2422)
3. Light DOM HTML structure differs before & after hydration (related to 1) – this is currently affecting our Storybook preview tab https://github.com/Esri/calcite-components/issues/2362

The biggest impact this would have is on form usage since it means that native form controls are no longer discoverable by light DOM forms. 

Fortunately, we only have a handful of scoped components:

* `calcite-inline-editable`
* `calcite-input`
* `calcite-label`
* `calcite-radio-button`

### Plan of action:

- [x] namespaced classes to help w/ 2.c - https://github.com/Esri/calcite-components/issues/2422  
- [x] assess impact on current form usage ([survey](https://survey123.arcgis.com/share/99cfa715e01843cf9bcd6ef060596228))
- [x] use shadow DOM on current scoped components (needs to happen early and tested early)
  - [x] `<calcite-label>` #3091
- [ ] add custom ESLint rule to disallow scope usage (out of scope)

Related: #2949 https://github.com/Esri/calcite-components/issues/1741

**Comment by driskull** [comment]
@jcfranco should we have issues to convert the components? I can start taking one of them.

**Comment by jcfranco** [comment]
@driskull To keep things simple, I think we can use this issue and have multiple PRs link to tasks/subtasks. cc @julio8a @caripizza 

On the conversion part, we're still waiting on results from the form-usage survey, but it'd be great if you can kick start a draft PR (shadow all the things!). 🥇

**Comment by driskull** [comment]
I think the biggest changes are going to be the logic for handling clicks on `calcite-label` elements.

If the native `label` and native `input` are now in shadow, clicking on a label won't focus or change the input.

In components like `calcite-radio-button`, we could listen for the event `calciteLabelFocus` and handle it. But what about native elements like `<calcite-label><input type="checkbox" /> test</calcite-label>`? The native label and native checkbox can't interact. Any ideas here?

**Comment by jcfranco** [comment]
@driskull Great points. Let's meet next week and draft a plan for this. The form-usage survey results should help streamline what we need to cover.

**Comment by jcfranco** [comment]
Quick update: @driskull and I met up last week and we'll work on some prototypes by the end of the week. We'll use this to gauge the strategy moving forward. 💪🏼

**Comment by jcfranco** [comment]
@benelan The work for `calcite-label` was installed and shipped in beta.66. One issue came up already: https://github.com/Esri/calcite-components/issues/2868

**Comment by benelan** [comment]
@jcfranco are you sure that's the right issue?

**Comment by jcfranco** [comment]
Updated, sorry about that (clipboard-history fingers working on their own this morning 😅 ).

**Comment by driskull** [comment]
Installed. Do we need to set for verification?

**Comment by jcfranco** [comment]
Here's a quick summary of the changes introduced by this work:

- Shadow DOM was enabled on the components mentioned in the description.
- Additional components are now form-associated (form submit, reset and input validity, if applicable), this is the current list:
  - `calcite-checkbox` 
  - `calcite-combobox` 
  - `calcite-input` 
  - `calcite-input-date-picker` 
  - `calcite-input-time-picker` 
  - `calcite-radio-button` 
  - `calcite-radio-button-group` 
  - `calcite-rating` 
  - `calcite-select` 
  - `calcite-slider` 
  - `calcite-switch`
- Each form-associated component implements a `disabled`, `name`, `required` and `value` properties
  - `start`/`end`/`startAsDate`/`endAsDate` deprecated in both `calcite-input-date-picker` & `calcite-date-picker`
  - added `required` props will not be documented until we determine how to handle them from a design perspective

For more details, please look at https://github.com/Esri/calcite-components/pull/2777

Next steps:

* Expose `calciteInputChange` – the `change` event is not composed and will no longer be available to end-users
* Extensively test updated components:
  * in plain setups
  * in frameworks
  * with forms
* Discuss how to tackle `required` behavior from a design perspective
* Update conventions accordingly

**Comment by jcfranco** [comment]
Adding to ☝, https://github.com/Esri/calcite-components/issues/3383 and https://github.com/Esri/calcite-components/issues/2949 will need to be tested again.

**Comment by benelan** [comment]
Seems like there haven't been any complaints and everything I've tried has worked. Is there anything specific I should do to verify this one? cc @jcfranco @driskull

**Comment by driskull** [comment]
I guess we can mark it verified and close 👍
